### DISCUSSION TITLE
is 與 == 相等？不相等？

### DISCUSSION THREAD

**Discussion by awkward-willy** [1101-GS4719-C/Python-Community-GS4719-C#438] [General]
今天上課時，老師有提到身分運算子的使用，其中is和is not是用以判斷兩物件是否為同樣的物件，即兩變數是否擁有相同的記憶體位置。
而由 Lab01 也可發現：
```python
a = 5
b = 5
a is b		# True
c = 5
d = 20
c is d		# False
```
這是不是就代表可以用is來判斷兩個數是否相同呢？
或許可以嘗試測試看看這幾種
```python
7777 is 7777	 # True, SyntaxWarning: "is" with a literal. Did you mean "=="?
a = 7777
b = 7777
a is b		 # False
c = 7777; d = 7777
c is d		 # True
```
！！！7777 難道不等於 7777 嗎？其實 Python 出於對效能的考量，在 [-5, 256] 的數值都採固定的記憶體位置，每次都存取於固定的記憶體位置，但超過這範圍後，每個值與變數都有其申請的記憶體位置。所以才會出現「同值不同址」

這件事情也跟老師給的心智圖中，int 是不可變（Mutable）的資料型態有關

所以若單純要比對值是否相同，還是乖乖用 == 吧

大家也可以試試看這些結果，看跟自己想的一不一樣（答案可能出乎意料之外）：
至於完整的詳細原因，小弟我也不完全清楚，知道的大大可以幫忙補充！
```python
a = 9487
b = 9453 + 34
a == b
a is b

c = 'Hello'
d = 'Hell' + 'o'
c == d
c is d

e = ['Hello', 'World']
f = ['Hello', 'World']
e == f
e is f

g = ['Hello', 'World']
h = g
g == h
g is h
```

[資料來源](https://clay-atlas.com/blog/2020/08/04/python-cn-equal-is-difference/)

**Comment by pekoranoinu** [comment]
感謝樓主的詳細解析
不然我一直以為is和==功用一樣

**Comment by ThomasMeng0515** [comment]
如果要看數值是否相等用 ==會比較適合 因為 is 是用來比對記憶體位址 也就是看變數的id函數結果是否相等。

**Reply by awkward-willy** [reply]
對ㄉ！我的內文就在分享這件事
標題好像讓你誤會為問題 抱歉抱歉w

**Reply by chia-kai-chang** [reply]
沒錯，比較運算子的比較對象是value，身分運算子的對象是記憶體id

**Comment by yi-1111** [comment]
原來如此，同值不一定同址，數值、記憶體位址大有學問

**Comment by xkllkx** [comment]
謝謝樓主的補充，這種數值類的bug真的很難抓，提早釐清觀念很重要

**Reply by chia-kai-chang** [reply]
大家都很認真的探究基本觀念，值得拍手鼓勵~

**Comment by sean1325418** [comment]
看來要比對值或是址的時候，工具要用對才可以

**Comment by chia-kai-chang** [comment]
> 其實 Python 出於對效能的考量，在 [-5, 256] 的數值都採固定的記憶體位置，每次都存取於固定的記憶體位置，但超過這範圍後，每個值與變數都有其申請的記憶體位置。所以才會出現「同值不同址」

沒錯~ 所以我在舉例時都要很小心~ 不要超過256~ 哈哈~ 😂

**Comment by Yuhsin000** [comment]
原來另有玄機，在[-5,256]這個區間用is 才會跟==一樣，因為比較的型態不同的關係，不知道哪個比較快

**Comment by chupochun** [comment]
原來如此，is跟==是有差別的，我一直以為兩者是相同概念

**Comment by Ku-Hao-Chen** [comment]
謝謝分享。

**Comment by PoooYen** [comment]
很實用的知識!!!

**Comment by d26488399** [comment]
感謝分享，終於懂了

**Comment by Sazmafuji** [comment]
這也太懂了吧

**Comment by sys511613** [comment]
感謝分享
另外要特別注意==中間不可以加空格，在任何地方都一樣

**Comment by minniewei** [comment]
感謝分享！！最近剛好也有這方面的問題

**Comment by treeann** [comment]
原來如此 一直以未差不多

**Comment by jcpan821** [comment]
感謝分享~~

**Comment by Ben6uo** [comment]
感謝分享 如果不說出來 都不知要卡多久

**Comment by alexthebigpp** [comment]
感謝分析，考前看到這篇剛好複習一下。

**Comment by 110409538** [comment]
感謝分享!!!

**Comment by bloodtrail0511** [comment]
感謝分享

**Comment by Lynn12378** [comment]
原來is跟==實際的差異在這，感謝分享!

**Comment by Jimmy911218** [comment]
感謝分享~小細節蠻需要注意的
